Letsencrypt

cseed opened the pull request:
Added back letsencrypt as (non-deployable) letsencrypt project.

Have gateway redirect the challenge to the letsencrypt service/pod.  letsencrypt pod includes kubectl and updates the secret.  Get cert for notebook.hail.is.  Current certs were generated with this.

Two more things to do:
 - run a process in gateway container to reload nginx when secrets change, and
 - set up renew cron job.

danking commented:
NB: GKE is up to k8s 1.10.7, so we can't use the 1.12 features yet.

cseed commented on letsencrypt/run-letsencrypt.sh:
```diff
@@ -0,0 +1,31 @@
+fi
+
+# cleanup
+kubectl delete pod --ignore-not-found=true letsencrypt
```
You can't delete the service because nginx restart will fail on a DNS resolution failure.

cseed commented:
> it seems to provide visibility into what happened during the last run of lets encrypt?

Yes.  As far as I know, certbot needs the previous config to do a renew (which I'm not doing yet).

 > I think the "sidecar" approach is simpler than this one (no extra nginx instance, no secrets, no service, no k8s secret creation privileges). We beef up the nginx pod to have a second container sharing a letsencrypt volume (which we've already defined in this PR)

You can't mount volumes to multiple pods.  You can't even mount volumes to the SAME pod if you want to do rolling updates (because the new instance can't launch because the old one is mounting the volume).  I think this means volumes for certs and web root are out.

volumes only work for replicated StatefulSets where you can take down one instance at a time for updates.

danking commented:
Hmm, actually GKE seems to vary in how quickly it gets new versions out.

It [supported k8s 1.11](https://cloud.google.com/kubernetes-engine/release-notes) for Early Access Partners (EAPs) within about a month. 1.12 was released a little less than a month ago.

The first generally available GKE k8s 1.10 release was May 15, and k8s 1.10 was released on March 26th, so that's less than two months from k8s to GKE.

cseed commented:
I addressed the comments, except for the following, which I plan to do as separate PRs:
 - put list of domains in a file,
 - least privileged: gateway-service-account (with certs read) and letsencrypt-sa (with certs update),
 - second process in gateway pod to bump nginx on cert change

cseed commented:
I moved the domains to a file.
